playbooks/openshift-certificates/private/restart.yml

```diff
@@ -0,0 +1,6 @@
+---
+- import_playbook: ../../openshift-etcd/private/restart.yml
```

Comment:
My thought about restarting them all at the end was to ensure that if there is a problem with verifying or replacing the certs on any component then there is a window to correct it before the entire cluster is restarted and begins using the new certs. If one component is restarted with new certs before the others it may not make valid connections until the rest of the cluster has new certs, this would especially be the case if a new CA is being provided.

Reply:
I would rather fail early on one component than have to wait until a bunch of malformed certs are deployed all over the place.

> If one component is restarted with new certs before the others it may not make valid connections until the rest of the cluster has new certs, this would especially be the case if a new CA is being provided.

Installing a new CA should be separated from this process.  We need to get out the new CA's first, then proceed with generating the certs / csrs.  Then we won't have this issue.

Reply:
> I would rather fail early on one component than have to wait until a bunch of malformed certs are deployed all over the place.

Hmm, so the proper order for the replacement playbook I think should now be:
1. Ensure all expected cluster certs and pending keys are present (and valid for usage)
2. Update etcd CA
3. Update etcd certs + keys + kubeconfigs
4. Restart etcd
5. Repeat 2 -> 4 for masters
6. Repeat 2 -> 4 for nodes

> Installing a new CA should be separated from this process. We need to get out the new CA's first, then proceed with generating the certs / csrs. Then we won't have this issue.

We can update the CAs before replacing the certs, but it does not make sense to do it before generation. After generation (csrs do not require a CA to generate) it may take some time before the admin has the issued certs and they may be signed by a CA the admin did not expect, so I would not want to commit to updating a CA until the replacement step.

Reply:
Need to update/add the new CA everywhere first.  If something needs to connect with etcd, it needs to trust the CA etcd's certs are signed with.  Just updating the CA on etcd doesn't help us here.

I don't think it's unreasonable to assume the CA should be determined before starting this process.

Reply:
> Need to update/add the new CA everywhere first. If something needs to connect with etcd, it needs to trust the CA etcd's certs are signed with. Just updating the CA on etcd doesn't help us here.

Ok, makes sense. Note that for any CA that the replacement playbook updates, I am creating a bundle consisting of the current CA and the replacement CA, and then linking to the bundle. So this means new bundles for etcd, frontproxy-ca, etc... I think this is the right thing to do just so at least momentarily the old CA is still trusted. Does this sound OK?

Reply:
I remember hearing a bundle proposed elsewhere and it seemed to be the consensus approach, so I think this is probably a good idea.

Reply:
I can see cases where the CA may not be eaxactly known in advance, but I think it is a restriction we can live by in Openshift.
However we need to have a way to remove the old CA at the end of the process.